### PR TITLE
feat: Migrate AccountTrackerController to BaseController v2

### LINKED ISSUE #25929: Migrate AccountTrackerController to BaseController v2
### What is this about?

Following the [Wallet Framework team's OKRs for Q3 2024](https://docs.google.com/document/d/1JLEzfUxHlT8lw8ntgMWG0vQb5BAATcrYZDj0wRB2ogI/edit#heading=h.kzzai3cfecro), we want to bring `AccountTracker` up to date with our latest controller patterns.

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

### Threat Modeling Framework

_No response_

### Acceptance Criteria

- A constant called `controllerName` exists which holds the name of the controller.
- A constant called `controllerMetadata` exists which holds the metadata for the state.
- The default state object is extracted to a function called `getDefaultAccountTrackerControllerState`.
- The class is renamed to `AccountTrackerController` and relocated to `app/scripts/controllers` along with its test file.
- `AccountTrackerController` inherits from `BaseController`.
  - The `AccountTrackerController` has a constructor which takes a partial `state` option (instead of `initState`) and a `messenger` option (instead of `controllerMessenger`), among other options.
  - The default state passed to ObservableStore (`initState`) is extracted and wrapped in a `getDefaultAccountTrackerControllerState` function.
  - The constructor obtains the current chain ID from the NetworkController (`NetworkController:getState` -> `selectedNetworkClientId` -> `NetworkController:getNetworkClientById` -> `configuration.chainId`) instead of taking a `getCurrentChainId` option.
  - The constructor uses the `NetworkController:getNetworkClientById` messenger action instead of taking a `getNetworkClientById` option.
  - The constructor uses the `PreferencesController:getState` messenger action instead of taking a `preferencesController` option.
  - The constructor uses the `OnboardingController:getState` messenger action and `OnboardingController:stateChange` messenger event instead of taking a `onboardingController` option.
  - `this.state` is used to access state instead of `this.store.getState`.
  - `this.update` is used to update state instead of `this.store.updateState`.
- Supporting types exist.
  - The `AccountTrackerControllerState` type exists and represents the current shape of the state object.
  - The `AccountTrackerControllerGetStateAction` and `AccountTrackerControllerStateChangeEvent` types exist.
  - The `AccountTrackerControllerActions` and `AccountTrackerControllerEvents` types exist.
  - The `AllowedActions` and `AllowedEvents` types exist.
  - The `AccountTrackerControllerMessenger` type exists and expects `NetworkController:getState`, `NetworkController:getNetworkClientById`, `PreferencesController:getState`, `OnboardingController:getState`, and `OnboardingController:stateChange` to be allowed.
- The tests are updated to follow suit.
  - `buildAccountTracker` is renamed to `buildAccountTrackerController`.
  - `buildAccountTrackerController` pre-registers action handlers using mock functions and returns those action handler mocks to be customized in tests.
  - `accountTrackerController.state` is used to access state instead of `accountTrackerController.store.getState()`.

### Stakeholder review needed before the work gets merged

- [X] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

- https://github.com/MetaMask/core/issues/1509

### PR DESCRIPTION
**Pull request by cryptodev-2s**
## **Description**

Migrate AccountTrackerController to BaseController v2

PS: Should be merged after the conversion to typescript is merged https://github.com/MetaMask/metamask-extension/pull/27231

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27258?quickstart=1)

## **Related issues**

Fixes: [#25929](https://github.com/MetaMask/metamask-extension/issues/25929)

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/controllers/account-tracker-controller.ts
Comment (MajorLift) [inline app/scripts/controllers/account-tracker-controller.ts:201]:
```suggestion
```

How would you feel about rebasing this branch on @kanthesha's https://github.com/MetaMask/metamask-extension/pull/27398 and replacing the direct usage of `this.#preferencesController` with messenger actions. It looks like all we need to achieve that is add `PreferencesControllerGetStateAction` to `AllowedActions`.

I think the necessary changes for removing this line belong in this PR since `AccountTrackerController` is downstream of `PreferencesController`. Ideally, we should have planned for `AccountTrackerController` to be blocked by `PreferencesController` for precisely this scenario.

File: app/scripts/controllers/account-tracker-controller.ts
Comment (MajorLift) [inline app/scripts/controllers/account-tracker-controller.ts:199]: Now that extension is using core's `NetworkController`, is it necessary to preserve the logic for pulling from the redux store using `getProviderConfig`? Is there anything keeping us from replacing this with a `NetworkController:getState` call that extracts `selectedNetworkClientId`?

File: app/scripts/controllers/account-tracker-controller.ts
Comment (cryptodev-2s) [reply app/scripts/controllers/account-tracker-controller.ts:201]: Although we already have the account tracker blocked by preference, I thought we could address this after the preferences upgrade is merged, considering the preferences controller hasn’t been integrated yet and for the sake of saving time.

Comment (MajorLift) [reply app/scripts/controllers/account-tracker-controller.ts:201]: The BaseControllerV2 upgrade feels incomplete when we're passing in a controller instance that can be replaced using the messenger.

If we can put a ticket on our board for immediately following up on this I'm OK with handling it separately.

But it looks like #27398 is closed to merging, and if it goes in first this change will become necessary regardless. I just tried the rebase and it isn't too painful.

Comment (cryptodev-2s) [reply app/scripts/controllers/account-tracker-controller.ts:201]: Yes, of course. I was planning to create a ticket for this if it gets merged before the preferences controller. Otherwise, I’ll simply rebase my work to use the messenger.

File: app/scripts/controllers/account-tracker-controller.ts
Comment (cryptodev-2s) [reply app/scripts/controllers/account-tracker-controller.ts]: Good catch updated

File: app/scripts/controllers/account-tracker-controller.ts
Comment (MajorLift) [reply app/scripts/controllers/account-tracker-controller.ts:199]: Now that we know we'll be migrating this to the core version that extends `StaticIntervalPollingController` we definitely will need to figure out a way to remove this line. 

I don't think we'd ever want a core controller relying on the redux store for stateful data. Seems like it would potentially introduce side effects and/or be a circular dependency?

Since the scope of this PR keeps the extension controller in use I guess we can handle this in the migration PR.

**Comment by MajorLift** [comment]
The sonarcloud quality gate warning is non-blocking for now, but that will change soon, and then someone will be stuck having to make up the coverage for the changes in the PR.

In this case I guess this isn't too problematic since we expect to remove the extension controller entirely.

File: app/scripts/controllers/account-tracker-controller.ts
Comment (cryptodev-2s) [reply app/scripts/controllers/account-tracker-controller.ts:199]: Yes, this will be handled in the migration PR. I’ll verify if it’s still needed, as it’s not part of the core implementation and is currently only being used to interact with the RPC using localhost as the hostname.

Comment (MajorLift) [reply app/scripts/controllers/account-tracker-controller.ts:199]: Plugging in the core controller into where the extension controller used to be will have the effect of removing this callback and its functionality. We should def keep an eye out in case it breaks anything, but it sounds like you have a good handle on what it's doing. :)

**Review by MajorLift — DISMISSED** [review]
Great work as always!

File: app/scripts/controllers/account-tracker-controller.ts
Comment (NiranjanaBinoy) [inline app/scripts/controllers/account-tracker-controller.ts:317]: nit: can we add some documentation for this method?

**Review by NiranjanaBinoy — APPROVED** [review]
LGTM

**Review by MajorLift — APPROVED** [review]

**Pull request merged**
